cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java

```diff
@@ -156,7 +161,101 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
       throw new RuntimeException(e);
     }
 
-    return new Pipeline(sourcePluginId, sinksInfo, transformInfos);
+    return new Pipeline(sourcePluginId, sinksInfo, transformInfos, connectionsMap);
+  }
+
+  @VisibleForTesting
+  static Map<String, List<String>> validateConnections(ETLConfig config) {
+    Map<String, List<String>> stageConnections = new HashMap<>();
+
+    // 1) basic validation, # of connections >= (source + #transform + sink - 1)
+    if (config.getConnections().size() < (config.getTransforms().size() + config.getSinks().size())) {
+      throw new IllegalArgumentException(
+        "Number of edges connecting the pipeline is less than the number of vertices, please check the connections");
+    }
+    // 2) connections end up in a sink and all sinks are covered also check there are no cycles.
+
+    // set containing names of stages.
+    Set<String> stageNamesFromConfig = new HashSet<>();
+
+    stageNamesFromConfig.add(config.getSource().getName());
+    for (ETLStage stage : config.getTransforms()) {
+      stageNamesFromConfig.add(stage.getName());
+    }
+    for (ETLStage stage : config.getSinks()) {
+      stageNamesFromConfig.add(stage.getName());
+    }
+
+    for (Connection connection : config.getConnections()) {
+      // check if this connections's from and to belong to actual stage
+      Preconditions.checkState(
+        stageNamesFromConfig.contains(connection.getFrom()),
+        String.format("Connection from : %s does not belong to an actual stage name, please check the config",
```

Comment:
message is a little unclear. How about "Invalid connection from '%s': '%s' is not a stage in the pipeline."